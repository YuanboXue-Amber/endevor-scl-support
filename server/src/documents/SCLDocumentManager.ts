import {
    TextDocument,
    TextDocumentItem,
    VersionedTextDocumentIdentifier,
    TextDocumentContentChangeEvent,
    TextDocumentPositionParams,
    CompletionItem,
    TextDocumentIdentifier,
    TextEdit,
    CompletionItemKind,
    Position, Range,
    Command,
    CodeLens,
    ExecuteCommandParams} from "vscode-languageserver";
import { SCLDocument } from './SCLDocument';
import { isNull, isNullOrUndefined } from "util";
import { isUnaryLike } from "@babel/types";
import { commands, executeSubmitSCL } from "../ExecuteCommandProvider";

export interface IDocumentSettings {
    maxNumberOfProblems: number;
    isREST: boolean;
}

interface ICapabilities {
    hasDiagnosticRelatedInformationCapability: boolean;
}

export const actionCompletion = [ "SET", "ADD", "UPDATE", "DELETE", "GENERATE",
"MOVE", "RETRIEVE", "SIGNIN", "TRANSFER", "APPROVE", "DENY", "BACKIN",
"BACKOUT", "CAST", "DEFINE", "EXECUTE", "RESET", "COMMIT", "LIST"];

/**
 * Manage all the opened documents in the client
 *
 * @export
 * @class SCLDocumentManager
 */
export class SCLDocumentManager {
    static config: IDocumentSettings = { maxNumberOfProblems: 1000, isREST: false };
    static capabilities: ICapabilities = { hasDiagnosticRelatedInformationCapability: false };

    static numberOfProblems = 0;

    // a map of document uri to document details
    documents: Map<string, SCLDocument> = new Map();

    constructor() {
        this.documents = new Map();
    }

    // NEVER USED
    openDocument(textDocumentItem: TextDocumentItem): SCLDocument {
        const textDocument: TextDocument = TextDocument.create(
            textDocumentItem.uri, textDocumentItem.languageId, textDocumentItem.version, textDocumentItem.text);
        const document: SCLDocument = new SCLDocument(textDocument);
        this.documents.set(textDocument.uri, document);

        // no need to refresh the diagnose number, since it is dealt with in SCLDocument.pushDiagnostic
        return document;
    }

    openOrChangeDocument(textDocument: TextDocument): SCLDocument {
        let document = this.documents.get(textDocument.uri);
        if (!document) {
            document = new SCLDocument(textDocument);
            this.documents.set(textDocument.uri, document);
            return document;
        }

        // reset the diagnose number
        let decreaseNum = 0;
        document.statements.forEach((statement) => {
            decreaseNum += statement.diagnostics.length;
        });
        SCLDocumentManager.numberOfProblems = SCLDocumentManager.numberOfProblems - decreaseNum;
        if (SCLDocumentManager.numberOfProblems < 0) {
            SCLDocumentManager.numberOfProblems = 0;
        }

        document.textDocument = textDocument;
        document.fullUpdate();
        this.documents.set(textDocument.uri, document);

        return document;
    }

    closeDocument(textDocumentUri: string): void {
        const document = this.documents.get(textDocumentUri);
        if (isNullOrUndefined(document)) {
            return;
        }

        // refresh the diagnose number
        let decreaseNum = 0;
        document.statements.forEach((statement) => {
            decreaseNum += statement.diagnostics.length;
        });
        SCLDocumentManager.numberOfProblems = SCLDocumentManager.numberOfProblems - decreaseNum;
        if (SCLDocumentManager.numberOfProblems < 0) {
            SCLDocumentManager.numberOfProblems = 0;
        }

        this.documents.delete(textDocumentUri);
    }

    getCompletionBySyntax(textDocumentPosition: TextDocumentPositionParams): CompletionItem[] {
        const document = this.documents.get(textDocumentPosition.textDocument.uri);
        if (!document) {
            throw new Error(`Cannot provide completion on unopened document: ${textDocumentPosition.textDocument.uri}`);
        }

        const tobeCompelteIndex = document.textDocument.offsetAt(textDocumentPosition.position);
        let completionItems: CompletionItem[] = [];

        for (const statement of document.statements) {
            if (tobeCompelteIndex >= statement.starti && tobeCompelteIndex <= statement.endi) {
                let i = -1;
                while (i < statement.tokens.length - 1) {
                    i ++;
                    const starti = statement.tokens[i].starti + statement.tokens[i].value.length;
                    const endi = (i+1) < statement.tokens.length ? statement.tokens[i+1].starti : statement.endi;

                    if (tobeCompelteIndex >= starti && tobeCompelteIndex <= endi) {
                        // to complete between tokens[i] and tokens[i+1] in scl
                        const values = statement.tokens[i].completionItems;
                        if (!isNullOrUndefined(values)) {
                            if (statement.tokens[i].value !== ".") {
                                completionItems = completionItems.concat(values);
                            } else {
                                for (const each of actionCompletion) {
                                    completionItems.push({
                                        label: each + " ",
                                        kind: CompletionItemKind.Function,
                                        documentation: "Endevor SCL keyword"
                                    });
                                }
                            }
                        }
                        return completionItems;
                    }
                }
            }
        }

        // to complete at the end of scl
        const statement = document.statements[document.statements.length-1];
        let values;
        if (!isNullOrUndefined(statement))
            values = statement.tokens[statement.tokens.length-1].completionItems;
        if (!isNullOrUndefined(values) && statement.tokens[statement.tokens.length-1].value !== ".") {
            completionItems = completionItems.concat(values);
        } else {
            // maybe a new scl is started
            for (const each of actionCompletion) {
                completionItems.push({
                    label: each + " ",
                    kind: CompletionItemKind.Function,
                    documentation: "Endevor SCL keyword"
                });
            }
        }
        return completionItems;
    }

    async formatDocument(textDocument: TextDocumentIdentifier): Promise<TextEdit[]> {
        const document = this.documents.get(textDocument.uri);
        if (!document) {
            throw new Error(`Cannot provide formatting on unopened document: ${textDocument.uri}`);
        }

        return document.formatDocument();
    }

    async computeCodeLenses(textDocument: TextDocumentIdentifier): Promise<CodeLens[]> {
        const document = this.documents.get(textDocument.uri);
        if (!document) {
            throw new Error(`Cannot provide codeLens on unopened document: ${textDocument.uri}`);
        }

        let codeLens: CodeLens[] = [];
        for (const statement of document.statements) {
            const range = {
                start: document.textDocument.positionAt(statement.starti),
                end: document.textDocument.positionAt(statement.starti+1),
            };
            codeLens = codeLens.concat(
                commands.map(command => ({
                    range,
                    command: Command.create(command.title, command.command, textDocument, range)
                }))
            );
        }
        return codeLens;
    }

    async executeCodeLens(params: ExecuteCommandParams) {
        const textDocument: TextDocumentIdentifier = (params.arguments as any[])[0];
        const range: Range = (params.arguments as any[])[1];
        const document = this.documents.get(textDocument.uri);
        if (!document) {
            throw new Error(`Cannot provide codeLens on unopened document: ${textDocument.uri}`);
        }
        return executeSubmitSCL(document, document.textDocument.offsetAt(range.start));
    }

}