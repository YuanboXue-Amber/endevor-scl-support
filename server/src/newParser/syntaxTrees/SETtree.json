{
    "keyword": "SET",
    "required": true,
    "next": [
        {
            "keyword": "FROM",
            "isFROM": true,
            "nogoback": true,
            "next": [
                {
                    "keyword": "FILe, DDName",
                    "next": [
                        {
                            "maxLen": 8,
                            "required": true,
                            "next": [
                                {
                                    "keyword": "."
                                }
                            ]
                        }
                    ]
                },
                {
                    "keyword": "DSName",
                    "next": [
                        {
                            "maxLen": 44,
                            "required": true,
                            "next": [
                                {
                                    "keyword": "."
                                }
                            ]
                        }
                    ]
                },
                {
                    "keyword": "SITE",
                    "next": [
                        {
                            "maxLen": 8,
                            "required": true,
                            "next": [
                                {
                                    "keyword": "ENVironment, SYStem, SUBSystem, TYPe",
                                    "next": [
                                        {
                                            "maxLen": 8,
                                            "required": true
                                        }
                                    ]
                                },
                                {
                                    "keyword": "STAge, STAge NUMber",
                                    "next": [
                                        {
                                            "maxLen": 1,
                                            "required": true
                                        }
                                    ]
                                },
                                {
                                    "keyword": "."
                                }
                            ]
                        }
                    ]
                },
                {
                    "keyword": "ENVironment, SYStem, SUBSystem, TYPe",
                    "next": [
                        {
                            "maxLen": 8,
                            "required": true,
                            "next": [
                                {
                                    "keyword": "."
                                }
                            ]
                        }
                    ]
                },
                {
                    "keyword": "STAge, STAge NUMber",
                    "next": [
                        {
                            "maxLen": 1,
                            "required": true,
                            "next": [
                                {
                                    "keyword": "."
                                }
                            ]
                        }
                    ]
                }
            ]
        },
        {
            "keyword": "TO",
            "isTO": true,
            "nogoback": true,
            "next": [
                {
                    "keyword": "SYSOut, C1PRINT, C1PRTVB",
                    "next": [
                        {
                            "keyword": "."
                        }
                    ]
                },
                {
                    "keyword": "FILe, DDName",
                    "next": [
                        {
                            "maxLen": 8,
                            "required": true,
                            "next": [
                                {
                                    "keyword": "."
                                }
                            ]
                        }
                    ]
                },
                {
                    "keyword": "DSName",
                    "next": [
                        {
                            "maxLen": 44,
                            "required": true,
                            "next": [
                                {
                                    "keyword": "MEMber",
                                    "next": [
                                        {
                                            "maxLen": 8,
                                            "required": true
                                        }
                                    ]
                                },
                                {
                                    "keyword": "."
                                }
                            ]
                        }
                    ]
                },
                {
                    "keyword": "ENVironment, SYStem, SUBSystem, TYPe",
                    "next": [
                        {
                            "maxLen": 8,
                            "required": true,
                            "next": [
                                {
                                    "keyword": "."
                                }
                            ]
                        }
                    ]
                },
                {
                    "keyword": "STAge, STAge NUMber",
                    "next": [
                        {
                            "maxLen": 1,
                            "required": true,
                            "next": [
                                {
                                    "keyword": "."
                                }
                            ]
                        }
                    ]
                }
            ]
        }
    ]
}