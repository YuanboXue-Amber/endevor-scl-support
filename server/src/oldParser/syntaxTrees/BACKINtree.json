{
    "value": "BACKIN",
    "type": "keyword",
    "requireNext": true,
    "next": [
        {
            "value": "PACKAGE",
            "type": "keyword",
            "requireNext": true,
            "next": [
                {
                    "value": "'packageId'",
                    "type": "value",
                    "maxLen": 16,
                    "next": [
                        {
                            "value": "STATEMENT",
                            "type": "keyword",
                            "requireNext": true,
                            "next": [
                                {
                                    "value": "NUMBER",
                                    "type": "keyword",
                                    "requireNext": true,
                                    "next": [
                                        {
                                            "value": "'1'",
                                            "type": "value",
                                            "requireNext": true,
                                            "next": [
                                                {
                                                    "value": "ELEMENT",
                                                    "type": "keyword",
                                                    "requireNext": true,
                                                    "next": [
                                                        {
                                                            "value": "'element'",
                                                            "type": "value",
                                                            "maxLen": 255,
                                                            "next": [
                                                                {
                                                                    "value": ".",
                                                                    "type": "eos",
                                                                    "next": []
                                                                }
                                                            ]
                                                        }
                                                    ]
                                                }
                                            ]
                                        }
                                    ]
                                },
                                {
                                    "value": "'1'",
                                    "type": "value",
                                    "requireNext": true,
                                    "next": [
                                        {
                                            "value": "ELEMENT",
                                            "type": "keyword",
                                            "requireNext": true,
                                            "next": [
                                                {
                                                    "value": "'element'",
                                                    "type": "value",
                                                    "maxLen": 255,
                                                    "next": [
                                                        {
                                                            "value": ".",
                                                            "type": "eos",
                                                            "next": []
                                                        }
                                                    ]
                                                }
                                            ]
                                        }
                                    ]
                                }
                            ]
                        },
                        {
                            "value": ".",
                            "type": "eos",
                            "next": []
                        }
                    ]
                }
            ]
        }
    ]
}