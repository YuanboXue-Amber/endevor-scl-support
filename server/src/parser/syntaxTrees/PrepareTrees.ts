import { Inode, ItreeNode } from './doc/Inode';
import path = require('path');
import fs = require('fs');
import { isNullOrUndefined } from "util";

export let SETtree: ItreeNode;
export let ADDtree: ItreeNode;
export let UPDATEtree: ItreeNode;
export let DELETEtree: ItreeNode;
export let GENERATEtree: ItreeNode;
export let MOVEtree: ItreeNode;

export function prepareTrees() {
    SETtree = composeTreeFromJSON(prepareJSON("SETtree.json"));
    ADDtree = composeTreeFromJSON(prepareJSON("ADDtree.json"));
    UPDATEtree = composeTreeFromJSON(prepareJSON("UPDATEtree.json"));
    DELETEtree = composeTreeFromJSON(prepareJSON("DELETEtree.json"));
    GENERATEtree = composeTreeFromJSON(prepareJSON("GENERATEtree.json"));
    MOVEtree = composeTreeFromJSON(prepareJSON("MOVEtree.json"));
}

export function composeTreeFromJSON(jsonRootNode: Inode):ItreeNode {
    const rootNode: ItreeNode = {
        value: jsonRootNode.value,
        type: jsonRootNode.type,
        requireNext: jsonRootNode.requireNext,
        children: []
    };

    const createChildTree = (jsonParentNode: Inode, jsonParentTreeNode: ItreeNode) => {
        let leftTreeNode: ItreeNode;
        jsonParentNode.next.forEach((childNode) => {
            let node: ItreeNode = {
                value: childNode.value,
                type: childNode.type,
                requireNext: childNode.requireNext,
                children: [],
                parent: jsonParentTreeNode
            };
            if (isNullOrUndefined(leftTreeNode)) {
                jsonParentTreeNode.children.push(node);
                createChildTree(childNode, node);
            } else {
                node.leftSibling = leftTreeNode;
                jsonParentTreeNode.children.push(node);
                createChildTree(childNode, node);
                leftTreeNode.rightSibling = node;
            }
            leftTreeNode = node;
        });
    };

    createChildTree(jsonRootNode, rootNode);
    return rootNode;
}

export function prepareJSON(jsonFileName: string): Inode {
    const jsonPath = path.resolve(__dirname, "../syntaxTrees/"+ jsonFileName);
    const jsonStr = fs.readFileSync(jsonPath, "UTF-8");
    return JSON.parse(jsonStr);
}


