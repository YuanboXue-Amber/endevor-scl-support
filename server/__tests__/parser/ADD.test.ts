import { Position } from 'vscode-languageserver';
import { SCLDocument } from '../../src/documents/SCLDocument';
import { prepareTrees } from '../../src/parser/syntaxTrees/PrepareTrees';

beforeAll(() => {
    prepareTrees();
});

describe("test1: Test 1 simple valid ADD ACTION", () => {
    it("Should have no error", async () => {
        let sourceTextDocument: any = {
            getText() {
                return " ADD  ELEMENT element FROM DSNAME test . ";
            },
            positionAt(offset: number): Position {
                expect(offset).toMatchSnapshot();
                const mockPosition: Position = {
                    line: 1,
                    character: 2
                };
                return mockPosition;
            },
            uri: "mocked/uri"
        };

        const document: SCLDocument = new SCLDocument(sourceTextDocument);
        for (const statement of document.statements) {
            expect(statement.diagnostics).toMatchSnapshot();
        }
    });
});

describe("test2: Test 1 simple INvalid ADD ACTION", () => {
    it("Should have error", async () => {
        let sourceTextDocument: any = {
            getText() {
                return " ADD  ELEMENT element FROM DSNAME test OPTION CCID . ";
            },
            positionAt(offset: number): Position {
                expect(offset).toMatchSnapshot();
                const mockPosition: Position = {
                    line: 1,
                    character: 2
                };
                return mockPosition;
            },
            uri: "mocked/uri"
        };

        const document: SCLDocument = new SCLDocument(sourceTextDocument);
        for (const statement of document.statements) {
            expect(statement.diagnostics).toMatchSnapshot();
        }
    });
});

describe("test3: Test 2 simple INvalid ADD ACTION", () => {
    it("Should have error", async () => {
        let sourceTextDocument: any = {
            getText() {
                return " ADD  ELEMENT element TO ENV OPTION \n new version 01 OPTION CCID .\n " +
                    "ADD  ELEMENT element TO ENV env1 SUBs sbs1 OPTION \n processor gro = 'group name'  .";
            },
            positionAt(offset: number): Position {
                expect(offset).toMatchSnapshot();
                const mockPosition: Position = {
                    line: 1,
                    character: 2
                };
                return mockPosition;
            },
            uri: "mocked/uri"
        };

        const document: SCLDocument = new SCLDocument(sourceTextDocument);
        for (const statement of document.statements) {
            expect(statement.diagnostics).toMatchSnapshot();
        }
    });
});

describe("test4: Test 1 valid ADD ACTION with multiple options", () => {
    it("Should have error", async () => {
        let sourceTextDocument: any = {
            getText() {
                return "ADD  ELEMENT element TO SUBs sbs1 ENV env1  OPTION \n processor gro = 'group name'  " +
                "OVErride SIGNOut CCID ccid . ";
            },
            positionAt(offset: number): Position {
                expect(offset).toMatchSnapshot();
                const mockPosition: Position = {
                    line: 1,
                    character: 2
                };
                return mockPosition;
            },
            uri: "mocked/uri"
        };

        const document: SCLDocument = new SCLDocument(sourceTextDocument);
        for (const statement of document.statements) {
            expect(statement.diagnostics).toMatchSnapshot();
        }
    });
});