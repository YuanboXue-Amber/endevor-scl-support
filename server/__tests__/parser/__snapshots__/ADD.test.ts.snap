// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test if the tree is prepared  1`] = `
Object {
  "next": Array [
    Object {
      "next": Array [
        Object {
          "next": Array [
            Object {
              "next": Array [
                Object {
                  "next": Array [
                    Object {
                      "next": Array [],
                      "type": "eos",
                      "value": ".",
                    },
                  ],
                  "type": "value",
                  "value": "element-name",
                },
              ],
              "requireNext": true,
              "type": "keyword",
              "value": "THROUGH",
            },
            Object {
              "next": Array [
                Object {
                  "next": Array [
                    Object {
                      "next": Array [],
                      "type": "eos",
                      "value": ".",
                    },
                  ],
                  "type": "value",
                  "value": "element-name",
                },
              ],
              "requireNext": true,
              "type": "keyword",
              "value": "THRU",
            },
            Object {
              "next": Array [
                Object {
                  "next": Array [
                    Object {
                      "next": Array [
                        Object {
                          "next": Array [],
                          "type": "eos",
                          "value": ".",
                        },
                      ],
                      "type": "value",
                      "value": "file-dd-name",
                    },
                  ],
                  "requireNext": true,
                  "type": "keyword",
                  "value": "FILE",
                },
                Object {
                  "next": Array [
                    Object {
                      "next": Array [
                        Object {
                          "next": Array [],
                          "type": "eos",
                          "value": ".",
                        },
                      ],
                      "type": "value",
                      "value": "dd-name",
                    },
                  ],
                  "requireNext": true,
                  "type": "keyword",
                  "value": "DDNAME",
                },
                Object {
                  "next": Array [
                    Object {
                      "next": Array [
                        Object {
                          "next": Array [],
                          "type": "eos",
                          "value": ".",
                        },
                        Object {
                          "next": Array [
                            Object {
                              "next": Array [
                                Object {
                                  "next": Array [],
                                  "type": "eos",
                                  "value": ".",
                                },
                              ],
                              "type": "value",
                              "value": "memeber-name",
                            },
                          ],
                          "requireNext": true,
                          "type": "keyword",
                          "value": "MEMBER",
                        },
                      ],
                      "type": "value",
                      "value": "data-set-name",
                    },
                  ],
                  "requireNext": true,
                  "type": "keyword",
                  "value": "DSNAME",
                },
                Object {
                  "next": Array [
                    Object {
                      "next": Array [
                        Object {
                          "next": Array [
                            Object {
                              "next": Array [
                                Object {
                                  "next": Array [],
                                  "type": "eos",
                                  "value": ".",
                                },
                              ],
                              "type": "value",
                              "value": "uss-filename",
                            },
                          ],
                          "requireNext": true,
                          "type": "keyword",
                          "value": "USSFILE",
                        },
                      ],
                      "requireNext": true,
                      "type": "value",
                      "value": "uss-path",
                    },
                  ],
                  "requireNext": true,
                  "type": "keyword",
                  "value": "PATH",
                },
              ],
              "requireNext": true,
              "type": "keyword",
              "value": "FROM",
            },
            Object {
              "next": Array [
                Object {
                  "next": Array [
                    Object {
                      "next": Array [
                        Object {
                          "next": Array [],
                          "type": "eos",
                          "value": ".",
                        },
                      ],
                      "type": "value",
                      "value": "environment-name",
                    },
                  ],
                  "requireNext": true,
                  "type": "keyword",
                  "value": "ENVIRONMENT",
                },
                Object {
                  "next": Array [
                    Object {
                      "next": Array [
                        Object {
                          "next": Array [],
                          "type": "eos",
                          "value": ".",
                        },
                      ],
                      "type": "value",
                      "value": "system-name",
                    },
                  ],
                  "requireNext": true,
                  "type": "keyword",
                  "value": "SYSTEM",
                },
                Object {
                  "next": Array [
                    Object {
                      "next": Array [
                        Object {
                          "next": Array [],
                          "type": "eos",
                          "value": ".",
                        },
                      ],
                      "type": "value",
                      "value": "subsystem-name",
                    },
                  ],
                  "requireNext": true,
                  "type": "keyword",
                  "value": "SUBSYSTEM",
                },
                Object {
                  "next": Array [
                    Object {
                      "next": Array [
                        Object {
                          "next": Array [],
                          "type": "eos",
                          "value": ".",
                        },
                      ],
                      "type": "value",
                      "value": "type-name",
                    },
                  ],
                  "requireNext": true,
                  "type": "keyword",
                  "value": "TYPE",
                },
              ],
              "requireNext": true,
              "type": "keyword",
              "value": "TO",
            },
            Object {
              "next": Array [
                Object {
                  "next": Array [
                    Object {
                      "next": Array [
                        Object {
                          "next": Array [],
                          "type": "eos",
                          "value": ".",
                        },
                      ],
                      "type": "value",
                      "value": "ccid-value",
                    },
                  ],
                  "requireNext": true,
                  "type": "keyword",
                  "value": "CCID",
                },
                Object {
                  "next": Array [
                    Object {
                      "next": Array [
                        Object {
                          "next": Array [],
                          "type": "eos",
                          "value": ".",
                        },
                      ],
                      "type": "value",
                      "value": "comment-value",
                    },
                  ],
                  "requireNext": true,
                  "type": "keyword",
                  "value": "COMMENT",
                },
                Object {
                  "next": Array [
                    Object {
                      "next": Array [
                        Object {
                          "next": Array [
                            Object {
                              "next": Array [],
                              "type": "eos",
                              "value": ".",
                            },
                          ],
                          "type": "value",
                          "value": "version-number",
                        },
                      ],
                      "requireNext": true,
                      "type": "keyword",
                      "value": "VERSION",
                    },
                  ],
                  "requireNext": true,
                  "type": "keyword",
                  "value": "NEW",
                },
                Object {
                  "next": Array [
                    Object {
                      "next": Array [
                        Object {
                          "next": Array [
                            Object {
                              "next": Array [],
                              "type": "eos",
                              "value": ".",
                            },
                          ],
                          "type": "keyword",
                          "value": "PRESENT",
                        },
                        Object {
                          "next": Array [],
                          "type": "eos",
                          "value": ".",
                        },
                      ],
                      "type": "keyword",
                      "value": "IF",
                    },
                    Object {
                      "next": Array [],
                      "type": "eos",
                      "value": ".",
                    },
                  ],
                  "type": "keyword",
                  "value": "UPDATE",
                },
                Object {
                  "next": Array [
                    Object {
                      "next": Array [
                        Object {
                          "next": Array [
                            Object {
                              "next": Array [],
                              "type": "eos",
                              "value": ".",
                            },
                          ],
                          "type": "keyword",
                          "value": "SOURCE",
                        },
                        Object {
                          "next": Array [],
                          "type": "eos",
                          "value": ".",
                        },
                      ],
                      "type": "keyword",
                      "value": "INPUT",
                    },
                    Object {
                      "next": Array [],
                      "type": "eos",
                      "value": ".",
                    },
                  ],
                  "type": "keyword",
                  "value": "DELETE",
                },
                Object {
                  "next": Array [
                    Object {
                      "next": Array [
                        Object {
                          "next": Array [],
                          "type": "eos",
                          "value": ".",
                        },
                      ],
                      "type": "keyword",
                      "value": "SIGNOUT",
                    },
                  ],
                  "requireNext": true,
                  "type": "keyword",
                  "value": "OVERRIDE",
                },
                Object {
                  "next": Array [
                    Object {
                      "next": Array [
                        Object {
                          "next": Array [
                            Object {
                              "next": Array [],
                              "type": "eos",
                              "value": ".",
                            },
                          ],
                          "type": "keyword",
                          "value": "PROCESSOR",
                        },
                      ],
                      "requireNext": true,
                      "type": "keyword",
                      "value": "GENERATE",
                    },
                  ],
                  "requireNext": true,
                  "type": "keyword",
                  "value": "BYPASS",
                },
                Object {
                  "next": Array [
                    Object {
                      "next": Array [
                        Object {
                          "next": Array [
                            Object {
                              "next": Array [
                                Object {
                                  "next": Array [],
                                  "type": "eos",
                                  "value": ".",
                                },
                              ],
                              "type": "value",
                              "value": "processor-group-name",
                            },
                          ],
                          "requireNext": true,
                          "type": "keyword",
                          "value": "EQUAL",
                        },
                        Object {
                          "next": Array [
                            Object {
                              "next": Array [
                                Object {
                                  "next": Array [],
                                  "type": "eos",
                                  "value": ".",
                                },
                              ],
                              "type": "value",
                              "value": "processor-group-name",
                            },
                          ],
                          "requireNext": true,
                          "type": "keyword",
                          "value": "EQUALSIGN",
                        },
                      ],
                      "requireNext": true,
                      "type": "keyword",
                      "value": "GROUP",
                    },
                  ],
                  "requireNext": true,
                  "type": "keyword",
                  "value": "PROCESSOR",
                },
                Object {
                  "next": Array [
                    Object {
                      "next": Array [],
                      "type": "eos",
                      "value": ".",
                    },
                    Object {
                      "next": Array [
                        Object {
                          "next": Array [
                            Object {
                              "next": Array [],
                              "type": "eos",
                              "value": ".",
                            },
                          ],
                          "type": "keyword",
                          "value": "NONE",
                        },
                        Object {
                          "next": Array [
                            Object {
                              "next": Array [],
                              "type": "eos",
                              "value": ".",
                            },
                          ],
                          "type": "keyword",
                          "value": "ALL",
                        },
                        Object {
                          "next": Array [
                            Object {
                              "next": Array [],
                              "type": "eos",
                              "value": ".",
                            },
                          ],
                          "type": "keyword",
                          "value": "SYSTEMS",
                        },
                        Object {
                          "next": Array [
                            Object {
                              "next": Array [],
                              "type": "eos",
                              "value": ".",
                            },
                          ],
                          "type": "keyword",
                          "value": "SUBSYSTEMS",
                        },
                      ],
                      "requireNext": true,
                      "type": "keyword",
                      "value": "SPAN",
                    },
                  ],
                  "type": "keyword",
                  "value": "AUTOGEN",
                },
              ],
              "requireNext": true,
              "type": "keyword",
              "value": "OPTION",
            },
            Object {
              "next": Array [],
              "type": "eos",
              "value": ".",
            },
          ],
          "type": "value",
          "value": "element-name",
        },
      ],
      "requireNext": true,
      "type": "keyword",
      "value": "ELEMENT",
    },
  ],
  "requireNext": true,
  "type": "keyword",
  "value": "ADD",
}
`;

exports[`test2: Test 1 simple INvalid ADD ACTION Should have error 1`] = `51`;

exports[`test2: Test 1 simple INvalid ADD ACTION Should have error 2`] = `52`;

exports[`test2: Test 1 simple INvalid ADD ACTION Should have error 3`] = `
Array [
  Object {
    "diagnostic": Object {
      "message": "Require a valid value to be specified",
      "range": Object {
        "end": Object {
          "character": 2,
          "line": 1,
        },
        "start": Object {
          "character": 2,
          "line": 1,
        },
      },
      "severity": 1,
      "source": "Endevor SCL extension",
    },
    "endi": 52,
    "starti": 51,
  },
]
`;

exports[`test3: Test 2 simple INvalid ADD ACTION Should have error 1`] = `38`;

exports[`test3: Test 2 simple INvalid ADD ACTION Should have error 2`] = `41`;

exports[`test3: Test 2 simple INvalid ADD ACTION Should have error 3`] = `101`;

exports[`test3: Test 2 simple INvalid ADD ACTION Should have error 4`] = `105`;

exports[`test3: Test 2 simple INvalid ADD ACTION Should have error 5`] = `120`;

exports[`test3: Test 2 simple INvalid ADD ACTION Should have error 6`] = `129`;

exports[`test3: Test 2 simple INvalid ADD ACTION Should have error 7`] = `130`;

exports[`test3: Test 2 simple INvalid ADD ACTION Should have error 8`] = `133`;

exports[`test3: Test 2 simple INvalid ADD ACTION Should have error 9`] = `
Array [
  Object {
    "diagnostic": Object {
      "message": "Invalid value specified",
      "range": Object {
        "end": Object {
          "character": 2,
          "line": 1,
        },
        "start": Object {
          "character": 2,
          "line": 1,
        },
      },
      "severity": 1,
      "source": "Endevor SCL extension",
    },
    "endi": 41,
    "starti": 38,
  },
]
`;

exports[`test3: Test 2 simple INvalid ADD ACTION Should have error 10`] = `
Array [
  Object {
    "diagnostic": Object {
      "message": "Keyword should be uppercased.
Lowercased keyword might cause the scl action to fail when submitted",
      "range": Object {
        "end": Object {
          "character": 2,
          "line": 1,
        },
        "start": Object {
          "character": 2,
          "line": 1,
        },
      },
      "severity": 2,
      "source": "Endevor SCL extension",
    },
    "endi": 105,
    "starti": 101,
  },
  Object {
    "diagnostic": Object {
      "message": "Keyword should be uppercased.
Lowercased keyword might cause the scl action to fail when submitted",
      "range": Object {
        "end": Object {
          "character": 2,
          "line": 1,
        },
        "start": Object {
          "character": 2,
          "line": 1,
        },
      },
      "severity": 2,
      "source": "Endevor SCL extension",
    },
    "endi": 129,
    "starti": 120,
  },
  Object {
    "diagnostic": Object {
      "message": "Keyword should be uppercased.
Lowercased keyword might cause the scl action to fail when submitted",
      "range": Object {
        "end": Object {
          "character": 2,
          "line": 1,
        },
        "start": Object {
          "character": 2,
          "line": 1,
        },
      },
      "severity": 2,
      "source": "Endevor SCL extension",
    },
    "endi": 133,
    "starti": 130,
  },
]
`;

exports[`test4: Test 1 valid ADD ACTION with multiple options Should have error 1`] = `24`;

exports[`test4: Test 1 valid ADD ACTION with multiple options Should have error 2`] = `28`;

exports[`test4: Test 1 valid ADD ACTION with multiple options Should have error 3`] = `53`;

exports[`test4: Test 1 valid ADD ACTION with multiple options Should have error 4`] = `62`;

exports[`test4: Test 1 valid ADD ACTION with multiple options Should have error 5`] = `63`;

exports[`test4: Test 1 valid ADD ACTION with multiple options Should have error 6`] = `66`;

exports[`test4: Test 1 valid ADD ACTION with multiple options Should have error 7`] = `83`;

exports[`test4: Test 1 valid ADD ACTION with multiple options Should have error 8`] = `91`;

exports[`test4: Test 1 valid ADD ACTION with multiple options Should have error 9`] = `92`;

exports[`test4: Test 1 valid ADD ACTION with multiple options Should have error 10`] = `99`;

exports[`test4: Test 1 valid ADD ACTION with multiple options Should have error 11`] = `
Array [
  Object {
    "diagnostic": Object {
      "message": "Keyword should be uppercased.
Lowercased keyword might cause the scl action to fail when submitted",
      "range": Object {
        "end": Object {
          "character": 2,
          "line": 1,
        },
        "start": Object {
          "character": 2,
          "line": 1,
        },
      },
      "severity": 2,
      "source": "Endevor SCL extension",
    },
    "endi": 28,
    "starti": 24,
  },
  Object {
    "diagnostic": Object {
      "message": "Keyword should be uppercased.
Lowercased keyword might cause the scl action to fail when submitted",
      "range": Object {
        "end": Object {
          "character": 2,
          "line": 1,
        },
        "start": Object {
          "character": 2,
          "line": 1,
        },
      },
      "severity": 2,
      "source": "Endevor SCL extension",
    },
    "endi": 62,
    "starti": 53,
  },
  Object {
    "diagnostic": Object {
      "message": "Keyword should be uppercased.
Lowercased keyword might cause the scl action to fail when submitted",
      "range": Object {
        "end": Object {
          "character": 2,
          "line": 1,
        },
        "start": Object {
          "character": 2,
          "line": 1,
        },
      },
      "severity": 2,
      "source": "Endevor SCL extension",
    },
    "endi": 66,
    "starti": 63,
  },
  Object {
    "diagnostic": Object {
      "message": "Keyword should be uppercased.
Lowercased keyword might cause the scl action to fail when submitted",
      "range": Object {
        "end": Object {
          "character": 2,
          "line": 1,
        },
        "start": Object {
          "character": 2,
          "line": 1,
        },
      },
      "severity": 2,
      "source": "Endevor SCL extension",
    },
    "endi": 91,
    "starti": 83,
  },
  Object {
    "diagnostic": Object {
      "message": "Keyword should be uppercased.
Lowercased keyword might cause the scl action to fail when submitted",
      "range": Object {
        "end": Object {
          "character": 2,
          "line": 1,
        },
        "start": Object {
          "character": 2,
          "line": 1,
        },
      },
      "severity": 2,
      "source": "Endevor SCL extension",
    },
    "endi": 99,
    "starti": 92,
  },
]
`;
